File: use-plugin/tools/runtime/LauncherMirrorProbe.java
import jacamo.infra.JaCaMoLauncher;
import cartago.*;

import java.net.URI;
import java.nio.file.*;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.tzi.use.plugins.jacamo.extraction.StaticProjectImporter;
import org.tzi.use.plugins.jacamo.mapping.*;
import org.tzi.use.plugins.jacamo.materialization.*;
import org.tzi.use.plugins.jacamo.runtime.*;
import org.tzi.use.plugins.jacamo.semantic.*;
import org.tzi.use.plugins.jacamo.trace.*;

/** Trusted derived control only. Launches real .jcm; never claims equivalence to Auction OS. */
public class LauncherMirrorProbe {
    static RuntimeMirrorService mirror;
    static MoiseRuntimeConnector moise;
    static CompositeRuntimeConnector composite;
    static RuntimeMutationEngine mutations;
    static final List<RuntimeEvent> events = new CopyOnWriteArrayList<>();
    static final List<Map<String,Object>> outcomes = new CopyOnWriteArrayList<>();
    static final List<RuntimeSnapshot> snapshots = new CopyOnWriteArrayList<>();
    static final List<String> checkpoints = new CopyOnWriteArrayList<>();
    static final List<Map<String,Object>> callbacks = new CopyOnWriteArrayList<>();

    public static void main(String[] args) throws Exception {
        Path module = Path.of(args[1]).toAbsolutePath(), output = Path.of(args[2]);
        var imported = new StaticProjectImporter().importProject(module.resolve("src/test/resources/auction/auction.jcm"));
        var semantic = imported.model();
        var mapping = new MappingLoader().loadCanonical(module);
        var structure = new TransformationPlanner().plan(semantic, mapping);
        var instances = new InstancePlanner().plan(semantic, mapping, structure);
        var generated = new TextBackend().generate("launcher_control", structure, instances);
        var direct = new DirectUseBackend().materialize(generated, instances);
        var trace = new TraceBuilder().build(semantic, mapping, structure, instances);
        var runtimeMapping = new RuntimeMappingLoader().loadDefault();
        String agent = id(semantic, MetamodelKind.Agent,"auctioneer");
        String artifact = id(semantic, MetamodelKind.Artifact,"auction1");
        var binding = new MoiseRuntimeBinding("auction_org", id(semantic,MetamodelKind.Organisation,"auction_org"),
                Map.of("auctioneer",agent), Map.of("auction_group",id(semantic,MetamodelKind.Group,"auction_group")),
                Map.of("auction_scheme",id(semantic,MetamodelKind.Scheme,"auction_scheme")));
        var artifactBinding = new CartagoArtifactBinding("/main/market","auction1",artifact,
                Map.of("open","open"),Map.of("placeBid","placeBid","closeAuction","closeAuction"));
        register(trace,agent,"jason:agent:auctioneer"); register(trace,agent,binding.agentRuntimeId("auctioneer"));
        register(trace,artifact,artifactBinding.runtimeSourceId());
        register(trace,binding.organisationSemanticId(),binding.organisationRuntimeId());
        binding.groups().forEach((k,v)->register(trace,v,binding.groupRuntimeId(k)));
        binding.schemes().forEach((k,v)->register(trace,v,binding.schemeRuntimeId(k)));
        JaCaMoLauncher launcher = new JaCaMoLauncher() {
            { // Same bootstrap as pinned JaCaMoLauncher.main, without interactive registries.
                runner = this;
                jason.runtime.RuntimeServicesFactory.set(new jacamo.infra.JaCaMoRuntimeServices(this));
            }
            protected java.io.InputStream getDefaultLogProperties() {
                return new java.io.ByteArrayInputStream("handlers=java.util.logging.ConsoleHandler\n.level=INFO\n".getBytes());
            }
            protected void startAgs() {
                var env = CartagoEnvironment.getInstance();
                moise = MoiseRuntimeConnector.forBoards("launcher-boards",binding,new MoiseBoardSnapshotSource("auction_org",
                    ()->ora4mas.nopl.GroupBoard.getGroupBoards().stream().filter(b->"auction_org".equals(b.getOEId())).toList(),
                    ()->ora4mas.nopl.SchemeBoard.getSchemeBoards().stream().filter(b->"auction_org".equals(b.getOEId())).toList()));
                var cartago = new CartagoRuntimeConnector("launcher-cartago",new RecordingAccess(env),List.of(artifactBinding));
                var jason = new JasonRuntimeConnector("launcher-jason",Map.of("auctioneer",getAg("auctioneer").getTS()),Map.of("auctioneer",agent));
                composite = new CompositeRuntimeConnector("launcher",List.of(jason,cartago,moise));
                mutations = new RuntimeMutationEngine(direct.system(),trace);
                mirror = new RuntimeMirrorService(composite,mutations,1024,new RuntimeEventObserver() {
                    public void eventReceived(RuntimeEvent e) { events.add(e); }
                    public void snapshotApplied(RuntimeSnapshot s) { snapshots.add(s); }
                    public void afterMutation(RuntimeEvent e, MutationResult r) {
                        var rule = runtimeMapping.select(e);
                        outcomes.add(Map.of("eventId",e.eventId(),"status",r.status().name(),
                            "diagnostic",String.valueOf(r.diagnostic()), "mappingRule",rule.id(),
                            "action",rule.action().name(), "runtimeSource",e.runtimeSourceId(),
                            "semanticSource",String.valueOf(e.semanticSourceId()),
                            "targets",trace.bySemanticId(e.semanticSourceId()).stream().map(t->t.targetUseId()).toList()));
                    }
                });
                mirror.connect(URI.create("jacamo://local/launcher-control"));
                require(mirror.state()==MirrorState.LIVE,"INITIAL_LIVE");
                super.startAgs();
            }
        };
        int exit=3;
        try {
            require(launcher.init(new String[]{args[0],"--no-net"})==0,"PARSE");
            launcher.create(); launcher.start();
            waitFor(()->ora4mas.nopl.GroupBoard.getGroupBoards().stream()
                    .anyMatch(b->"auction_org".equals(b.getOEId()) && b.getGrpState().hasPlayer("auctioneer","auctioneer")),"ROLE_READY");
            require(moise.discoveredGroups().equals(List.of("auction_group")),"EXACT_GROUP");
            require(moise.discoveredSchemes().equals(List.of("auction_scheme")),"EXACT_SCHEME");
            moise.pollChanges(); mirror.awaitIdle(Duration.ofSeconds(5));
            // Input triggers a checked-in probe-only AgentSpeak plan. Artifact calls are performed by Jason.
            launcher.getAg("auctioneer").getTS().getC().addAchvGoal(jason.asSyntax.ASSyntax.createLiteral("probe"),jason.asSemantics.Intention.EmptyInt);
            launcher.getAg("auctioneer").wake();
            waitFor(()->launcher.getAg("auctioneer").getTS().getAg().getBB()
                    .contains(jason.asSyntax.ASSyntax.createLiteral("probe_done"))!=null,"AGENT_SCENARIO_DONE");
            require(ora4mas.nopl.SchemeBoard.getSchemeBoards().stream().anyMatch(b->
                "auction_org".equals(b.getOEId()) && "auction_scheme".equals(b.getArtId())
                && b.getSchState().hasPlayer("auctioneer","run_auction")
                && b.getSchState().isSatisfied(b.getSpec().getGoal("sell_item"))),"MISSION_AND_GOAL_READY");
            moise.pollChanges();
            mirror.awaitIdle(Duration.ofSeconds(5));
            launcher.getAg("auctioneer").getTS().getC().addAchvGoal(jason.asSyntax.ASSyntax.createLiteral("probe_failure"),jason.asSemantics.Intention.EmptyInt);
            launcher.getAg("auctioneer").wake();
            waitFor(()->launcher.getAg("auctioneer").getTS().getAg().getBB()
                    .contains(jason.asSyntax.ASSyntax.createLiteral("failure_seen"))!=null,"AGENT_FAILURE_HANDLED");
            mirror.awaitIdle(Duration.ofSeconds(5));
            require(events.stream().anyMatch(e->e.kind()==RuntimeEventKind.OP_FAIL),"OP_FAIL");
            require(events.stream().anyMatch(e->e.kind()==RuntimeEventKind.OP_ENTER),"OP_ENTER");
            require(events.stream().anyMatch(e->e.kind()==RuntimeEventKind.OP_EXIT),"OP_EXIT");
            require(events.stream().anyMatch(e->e.kind()==RuntimeEventKind.OBS_PROPERTY_CHANGED),"PROPERTY_CHANGE");
            require(events.stream().anyMatch(e->e.kind()==RuntimeEventKind.ROLE_ADOPTED),"ROLE_DELTA");
            require(events.stream().anyMatch(e->e.kind()==RuntimeEventKind.MISSION_COMMITTED),"MISSION_DELTA");
            require(events.stream().anyMatch(e->e.kind()==RuntimeEventKind.SCHEME_STATE_CHANGED
                && "satisfied".equals(e.payload().get("state"))),"GOAL_DELTA");
            require(!mirror.checkDrift(DriftResyncPolicy.REPORT_ONLY).drifted(),"INITIAL_DRIFT");
            mirror.disconnect();
            int before=events.size();
            // Exact trusted fixture stimulus while disconnected; no plugin-initiated business action.
            launcher.getAg("auctioneer").getTS().getC().addAchvGoal(jason.asSyntax.ASSyntax.createLiteral("probe_disconnected"),jason.asSemantics.Intention.EmptyInt);
            launcher.getAg("auctioneer").wake();
            waitFor(()->launcher.getAg("auctioneer").getTS().getAg().getBB()
                    .contains(jason.asSyntax.ASSyntax.createLiteral("disconnected_done"))!=null,"DISCONNECTED_MUTATION");
            require(events.size()==before,"DISCONNECTED_LISTENER");
            mirror.connect(URI.create("jacamo://local/launcher-control"));
            require(mirror.state()==MirrorState.LIVE,"RECONNECT_LIVE");
            require(!mirror.checkDrift(DriftResyncPolicy.REPORT_ONLY).drifted(),"RECONNECT_DRIFT");
            var metrics=mirror.metrics();
            require(metrics.failed()==0 && metrics.rejected()==0 && metrics.dropped()==0,"QUEUE");
            require(outcomes.stream().allMatch(o->"APPLIED".equals(o.get("status"))),"ALL_MUTATIONS_ACCOUNTED");
            Files.createDirectories(output);
            Files.writeString(output.resolve("model.use"),generated.useModel());
            Files.writeString(output.resolve("initial-state.cmd"),generated.initialCommands());
            new RuntimeEventCodec().writeEvents(output.resolve("runtime-events.json"),events);
            var json=new ObjectMapper();
            json.writerWithDefaultPrettyPrinter().writeValue(output.resolve("outcomes.json").toFile(),outcomes);
            json.writerWithDefaultPrettyPrinter().writeValue(output.resolve("checkpoints.json").toFile(),checkpoints);
            json.writerWithDefaultPrettyPrinter().writeValue(output.resolve("raw-callbacks.json").toFile(),callbacks);
            List<RuntimeEvent> initial = snapshots.stream().flatMap(s->s.mutations().stream()).toList();
            new RuntimeEventCodec().writeEvents(output.resolve("snapshot-events.json"),initial);
            new org.tzi.use.plugins.jacamo.trace.TraceStore().write(output.resolve("trace.json"),trace);
            json.writerWithDefaultPrettyPrinter().writeValue(output.resolve("summary.json").toFile(),Map.of(
                "status","SUPPORTED_SUBSET_COMPLETE","originalAuctionEquivalent",false,
                "launcher","JaCaMo 1.3.0 .jcm","operationDriver","actual Jason agent",
                "organisationSource","actual launcher GroupBoard/SchemeBoard; no independent OE",
                "snapshots",snapshots.size(),"events",events.size(),"unexplainedDrift",0,
                "scope","Derived control only; original plan/deadline semantics remain unproven"));
            System.out.println("PHASE20_LAUNCHER_MIRROR_CONTROL_PASS");
            exit=0;
        } catch(Throwable e) {
            e.printStackTrace();
            var diagnostics = new LinkedHashMap<String,Object>();
            diagnostics.put("failure",e.toString()); diagnostics.put("checkpoints",checkpoints);
            diagnostics.put("mirrorState",mirror==null ? "NOT_ATTACHED" : mirror.state().name());
            diagnostics.put("groups",ora4mas.nopl.GroupBoard.getGroupBoards().stream()
                .map(b->b.getOEId()+"/"+b.getArtId()+":"+b.getGrpState()).toList());
            diagnostics.put("schemes",ora4mas.nopl.SchemeBoard.getSchemeBoards().stream()
                .map(b->b.getOEId()+"/"+b.getArtId()+":"+b.getSchState()).toList());
            diagnostics.put("agentBeliefs",launcher.getAg("auctioneer")==null ? "ABSENT"
                : launcher.getAg("auctioneer").getTS().getAg().getBB().toString());
            new ObjectMapper().writerWithDefaultPrettyPrinter().writeValue(output.resolve("failure.json").toFile(),diagnostics);
        }
        finally {
            try { if(mirror!=null) { mirror.close(); require(composite.state()==ConnectorState.DISCONNECTED,"CONNECTOR_CLEANUP"); } }
            catch(Throwable e) { e.printStackTrace(); exit=3; }
            try { launcher.finish(0,false,0); } catch(Throwable e) { e.printStackTrace(); exit=3; }
        }
        new ObjectMapper().writerWithDefaultPrettyPrinter().writeValue(output.resolve("checkpoints.json").toFile(),checkpoints);
        System.exit(exit); // Isolated process exit is containment, not proof of upstream thread quiescence.
    }
    static String id(JaCaMoSemanticModel m,MetamodelKind k,String name) {
        return m.elements().stream().filter(e->e.kind()==k && e.name().equals(name)).findFirst().orElseThrow().id().value();
    }
    static void register(TraceIndex trace,String semantic,String runtime) {
        var record=trace.bySemanticId(semantic).stream().filter(r->"OBJECT".equals(r.targetKind())).findFirst().orElseThrow();
        trace.registerRuntimeKey(record.traceId(),runtime);
    }
    static void require(boolean pass,String code) {
        if(!pass)throw new IllegalStateException("PHASE20_CONTROL_"+code);
        checkpoints.add(code);
    }
    /** Decorates the one connector-owned logger; does not register a second listener. */
    static final class RecordingAccess implements CartagoRuntimeAccess {
        final OfficialCartagoRuntimeAccess delegate;
        final Map<ICartagoLogger,ICartagoLogger> wrappers = new IdentityHashMap<>();
        RecordingAccess(CartagoEnvironment env) { delegate = new OfficialCartagoRuntimeAccess(env); }
        public ICartagoController controller(String workspace) throws CartagoException { return delegate.controller(workspace); }
        public synchronized void registerLogger(String workspace, ICartagoLogger logger) throws CartagoException {
            var wrapper = (ICartagoLogger)java.lang.reflect.Proxy.newProxyInstance(
                ICartagoLogger.class.getClassLoader(),new Class<?>[]{ICartagoLogger.class},(proxy,method,args)->{
                    if(method.getDeclaringClass()==Object.class) return switch(method.getName()) {
                        case "equals" -> proxy==args[0]; case "hashCode" -> System.identityHashCode(proxy);
                        default -> "Phase20RecordingLogger";
                    };
                    callbacks.add(Map.of("category",method.getName(),"workspace",workspace,
                        "receivedAt",java.time.Instant.now().toString(),"arguments",
                        args==null ? List.of() : Arrays.stream(args).map(LauncherMirrorProbe::rawValue).toList()));
                    try { return method.invoke(logger,args); }
                    catch(java.lang.reflect.InvocationTargetException e) { throw e.getCause(); }
                });
            require(!wrappers.containsKey(logger),"LOGGER_UNIQUE");
            delegate.registerLogger(workspace,wrapper); wrappers.put(logger,wrapper);
        }
        public synchronized void unregisterLogger(String workspace, ICartagoLogger logger) throws CartagoException {
            var wrapper = wrappers.get(logger);
            require(wrapper!=null,"LOGGER_REGISTERED");
            delegate.unregisterLogger(workspace,wrapper); wrappers.remove(logger);
        }
    }
    static Object rawValue(Object value) {
        if(value==null) return "null";
        if(value instanceof Number || value instanceof Boolean || value instanceof String) return value;
        if(value instanceof Object[] array) return Arrays.stream(array).map(LauncherMirrorProbe::rawValue).toList();
        if(value instanceof Op o) return Map.of("name",o.getName(),"parameters",rawValue(o.getParamValues()));
        if(value instanceof ArtifactObsProperty p) return Map.of("name",p.getName(),"values",rawValue(p.getValues()));
        if(value instanceof ArtifactId a) return Map.of("name",a.getName(),"id",a.getId().toString(),"workspace",a.getWorkspaceId().getFullName());
        if(value instanceof AgentId a) return Map.of("name",a.getAgentName(),"id",a.getGlobalId());
        if(value instanceof OpId o) return Map.of("id",o.getId(),"operation",o.getOpName(),"artifact",rawValue(o.getArtifactId()),"agent",rawValue(o.getAgentBodyId()));
        return value.toString();
    }
    static void waitFor(java.util.function.BooleanSupplier condition,String code) throws Exception {
        long end=System.nanoTime()+Duration.ofSeconds(10).toNanos();
        while(!condition.getAsBoolean() && System.nanoTime()<end)Thread.sleep(10);
        require(condition.getAsBoolean(),code);
    }
}

File: use-plugin/tools/runtime/launcher_mirror_probe.py
"""Run the trusted, explicitly non-equivalent launcher mirror control after launcher_probe.py.
Requires mvn -pl use-plugin test-compile and the existing pinned launcher classpath.
Never executes an arbitrary user-supplied project. Parent timeout/reap is mandatory.
"""
import hashlib, json, os, pathlib, subprocess, sys

module = pathlib.Path(__file__).resolve().parents[2]
negative = '--negative-readiness' in sys.argv
output = module / ('target/phase20-launcher-negative' if negative else 'target/phase20-launcher-mirror')
output.mkdir(parents=True, exist_ok=True)
for stale in ('summary.json', 'failure.json'):
    (output/stale).unlink(missing_ok=True)
repository = pathlib.Path(os.environ.get('MAVEN_REPO_LOCAL', pathlib.Path.home()/'.m2/repository'))
launcher = repository/'org/jacamo/jacamo/1.3.0/jacamo-1.3.0.jar'
assert launcher.is_file(), 'PHASE20_PINNED_DEPENDENCY_MISSING'
classpath = (module/'target/phase20-component-classpath.txt').read_text().strip()
classpath = os.pathsep.join([str(module/'target/classes'), classpath, str(launcher),
    str(repository/'org/jacamo/jaca/3.1/jaca-3.1.jar'), str(repository/'org/jacamo/npl/0.6/npl-0.6.jar'),
    str(module/'target/phase20-launcher')])
project = (module/'target/phase20-launcher/builder-control.jcm').read_text()
original_asl_root = (module/'src/test/resources/auction/src/agt').as_posix()
project = project.replace('auctioneer.asl','probe.asl').replace('goals: start','')
project = project.replace(original_asl_root, output.as_posix())
project = project.replace('org-path: "'+(module/'target/phase20-launcher').as_posix()+'"',
                          'org-path: "'+output.as_posix()+'"')
# Explicit control policy: permission to execute one mission, no source deadline/plan equivalence.
control_os = (module/'target/phase20-launcher/builder-control.xml').read_text()
control_os = control_os.replace('<normative-specification/>', '<normative-specification><norm id="control_permission" role="auctioneer" mission="run_auction" type="permission"/></normative-specification>')
(output/'builder-control.xml').write_text(control_os)
if negative:
    project = project.replace('roles: auctioneer in auction_group','').replace('players: auctioneer auctioneer','')
(output/'probe.jcm').write_text(project)
(output/'probe.asl').write_text('''// Trusted diagnostic control, not the original Auction agent semantics.
{ include("$jacamoJar/templates/common-cartago.asl") }
{ include("$jacamoJar/templates/common-moise.asl") }
+!probe <- commitMission(run_auction)[artifact_name(auction_scheme),wsp(auction_org)];
           placeBid("item1",10)[artifact_name(auction1),wsp(market)];
           closeAuction[artifact_name(auction1),wsp(market)];
           goalAchieved(sell_item)[artifact_name(auction_scheme),wsp(auction_org)]; +probe_done.
+!probe_failure <- placeBid("item1",0)[artifact_name(auction1),wsp(market)].
-!probe_failure <- +failure_seen.
+!probe_disconnected <- removeOpen[artifact_name(auction1),wsp(market)]; +disconnected_done.
''')
java = pathlib.Path(os.environ['JAVA_HOME'])/'bin/java.exe'
log = output/'probe.log'
command = [str(java),'-Djava.awt.headless=true','-cp',classpath,
           str(module/'tools/runtime/LauncherMirrorProbe.java'),str(output/'probe.jcm'),str(module),str(output)]
with log.open('w',encoding='utf-8') as stream:
    try:
        result=subprocess.run(command,cwd=module,stdout=stream,stderr=subprocess.STDOUT,timeout=45)
    except subprocess.TimeoutExpired:
        (output/'failure.json').write_text(json.dumps({'failure':'PHASE20_PROCESS_TIMEOUT', 'timeoutSeconds':45}))
        raise SystemExit('PHASE20_PROCESS_TIMEOUT: child killed and reaped; inspect '+str(log))
text=log.read_text(encoding='utf-8')
print(text)
if negative:
    assert result.returncode != 0 and 'PHASE20_CONTROL_ROLE_READY' in text, 'READINESS_NEGATIVE_WRONG_FAILURE'
    assert 'PHASE20_LAUNCHER_MIRROR_CONTROL_PASS' not in text, 'FALSE_POSITIVE_READINESS'
    failure=json.loads((output/'failure.json').read_text())
    assert failure['failure'].endswith('PHASE20_CONTROL_ROLE_READY'), 'READINESS_NEGATIVE_WRONG_GATE'
    assert failure['mirrorState']=='LIVE' and failure['groups'] and failure['schemes'], 'NEGATIVE_BOARD_PRECONDITION_MISSING'
    assert 'INITIAL_LIVE' in failure['checkpoints'] and 'AGENT_SCENARIO_DONE' not in failure['checkpoints']
    print('PHASE20_READINESS_NEGATIVE_PASS: existing boards without role rejected')
    raise SystemExit(0)
if result.returncode or 'PHASE20_LAUNCHER_MIRROR_CONTROL_PASS' not in text:
    raise SystemExit('PHASE20_LAUNCHER_MIRROR_CONTROL_FAILED')
summary=json.loads((output/'summary.json').read_text())
events=json.loads((output/'runtime-events.json').read_text())['events']
outcomes=json.loads((output/'outcomes.json').read_text())
callbacks=json.loads((output/'raw-callbacks.json').read_text())
checkpoints=json.loads((output/'checkpoints.json').read_text())
assert 'CONNECTOR_CLEANUP' in checkpoints, 'CONNECTOR_CLEANUP_NOT_PROVEN'
assert len(events)==len(outcomes) and {e['eventId'] for e in events}=={o['eventId'] for o in outcomes}, 'UNACCOUNTED_EVENT'
assert all(o['status']=='APPLIED' and o['mappingRule'] and o['targets'] for o in outcomes), 'UNRESOLVED_MUTATION'
assert all(e['timestamp'] and e['runtimeSourceId'] and e['semanticSourceId'] for e in events), 'TRACE_IDENTITY_MISSING'
sequences=[e['sequence'] for e in events]
assert sequences==sorted(set(sequences)), 'EVENT_ORDER_OR_DUPLICATE'
operations={}
for event in events:
    if event['kind'] in ('OP_ENTER','OP_EXIT','OP_FAIL'):
        assert event['correlationId'], 'MISSING_OPERATION_CORRELATION'
        operations.setdefault(event['correlationId'], []).append(event)
assert len(operations)==3, 'CONTROL_OPERATION_COUNT'
for lifecycle in operations.values():
    assert [e['kind'] for e in lifecycle] in (['OP_ENTER','OP_EXIT'],['OP_ENTER','OP_FAIL']), 'OPERATION_LIFECYCLE_MISMATCH'
    assert len({e['runtimeSourceId'] for e in lifecycle})==1, 'OPERATION_IDENTITY_MISMATCH'
    assert all(e['payload']['agent']=='auctioneer' for e in lifecycle), 'NOT_ACTUAL_AGENT_OPERATION'
properties=[e for e in events if e['kind']=='OBS_PROPERTY_CHANGED']
assert len(properties)==1 and properties[0]['payload']['property']=='open' and properties[0]['payload']['value'] is False, 'PROPERTY_DELTA_MISMATCH'
close=next(es for es in operations.values() if es[0]['payload']['operation']=='closeAuction')
assert close[0]['sequence'] < properties[0]['sequence'] < close[1]['sequence'], 'PROPERTY_TIMING_MISMATCH'
assert any(es[0]['payload']['arguments']==['item1',10] and es[-1]['kind']=='OP_EXIT' for es in operations.values()), 'NO_STATE_CHANGE_OPERATION_MISSING'
assert any(es[0]['payload']['arguments']==['item1',0] and es[-1]['kind']=='OP_FAIL' for es in operations.values()), 'FAILURE_OPERATION_MISSING'
assert {'opRequested','opStarted','opCompleted','opFailed','newPercept'} <= {c['category'] for c in callbacks}, 'RAW_CALLBACK_COVERAGE'
summary['traceContractChecks']='PASS: exact identity, complete outcomes, ordered unique sequence, paired operation correlations, arguments, property timing, raw callback categories, cleanup'
summary['observedEventKinds']=sorted({e['kind'] for e in events})
summary['rawCallbackCategories']=sorted({c['category'] for c in callbacks})
enum=(module/'src/main/java/org/tzi/use/plugins/jacamo/runtime/RuntimeEventKind.java').read_text()
known={name.strip() for name in enum.split('{',1)[1].rsplit('}',1)[0].split(',')}
assert set(summary['observedEventKinds']) <= known, 'UNKNOWN_EVENT_KIND'
summary['unobservedEventKinds']=sorted(known-set(summary['observedEventKinds']))
summary['unobservedMeaning']='Not exercised by this scenario; not evidence of API impossibility'
summary['processExitCode']=result.returncode
summary['sourceRevision']=subprocess.check_output(['git','rev-parse','HEAD'],cwd=module,text=True).strip()
summary['workingTreeStatus']=subprocess.check_output(['git','status','--porcelain'],cwd=module,text=True).strip()
summary['processContainment']='child returned and reaped; not a claim of upstream thread-level quiescence'
summary['inputSha256']={str(p.relative_to(module)):hashlib.sha256(p.read_bytes()).hexdigest() for p in
    [module/'tools/runtime/LauncherMirrorProbe.java',pathlib.Path(__file__).resolve(),
     module/'src/main/java/org/tzi/use/plugins/jacamo/runtime/MoiseBoardSnapshotSource.java',
     module/'src/main/java/org/tzi/use/plugins/jacamo/runtime/MoiseRuntimeConnector.java']}
summary['sha256']={p.name:hashlib.sha256(p.read_bytes()).hexdigest() for p in sorted(output.iterdir()) if p.is_file() and p.name!='summary.json'}
(output/'summary.json').write_text(json.dumps(summary,indent=2)+'\n')
